src/Resolution.ts

```diff
@@ -552,29 +405,34 @@ export default class Resolution {
 
   private getDnsRecordKeys(types: DnsRecordType[]): string[] {
     const records = ['dns.ttl'];
-    types.forEach(type => {
+    types.forEach((type) => {
       records.push(`dns.${type}`);
       records.push(`dns.${type}.ttl`);
     });
     return records;
   }
 
-  private async getPreferableNewRecord(domain: string, newRecord: string, oldRecord: string): Promise<string> {
-    const records = await this.records(domain, [newRecord, oldRecord]) as Record<string, string>;
-    return NamingService.ensureRecordPresence(domain, newRecord, records[newRecord] || records[oldRecord]);
+  private async getPreferableNewRecord(
+    domain: string,
+    newRecord: string,
+    oldRecord: string,
+  ): Promise<string> {
+    const records = (await this.records(domain, [
+      newRecord,
+      oldRecord,
+    ]));
+    if (!records[newRecord] && !records[oldRecord]) {
+      throw new ResolutionError(ResolutionErrorCode.RecordNotFound, {
+        recordName: newRecord,
+        domain: domain,
+      });
+    }
+    return (records[newRecord] || records[oldRecord]);
   }
 
   private getNamingMethod(domain: string): NamingService | undefined {
-    return this.getResolutionMethods().find(method =>
-      method.isSupportedDomain(domain),
-    );
-  }
-
-  private getResolutionMethods(): NamingService[] {
-    return (this.blockchain
-      ? ([this.ens, this.zns, this.cns] as NamingService[])
-      : ([this.api] as NamingService[])
-    ).filter(v => v);
+    const lastWord = domain.split(".").pop() || '';
```

Comment:
@JohnnyJumper Detecting service name seems quite important logic. It's better to move it into shared function and reuse in UdAPI class